File: sdk/python/src/openlit/instrumentation/strands/processor.py
"""Strands Agents SpanProcessor.

Enriches Strands' native OTel spans with OpenLIT-specific attributes,
extracts content from span events into span attributes, emits
``gen_ai.client.inference.operation.details`` log events for ``chat``
spans (matching the OpenAI instrumentation pattern), and records
OpenLIT metrics.

Provider-level chat spans (OpenAI, Anthropic, etc.) are suppressed
when they occur inside a Strands ``chat`` span via the shared
``_framework_llm_span_active`` ContextVar.
"""

import json
import logging
import threading
import types

from opentelemetry.sdk.resources import (
    SERVICE_NAME,
    TELEMETRY_SDK_NAME,
    DEPLOYMENT_ENVIRONMENT,
)
from opentelemetry.sdk.trace import SpanProcessor

from openlit.__helpers import (
    get_chat_model_cost,
    set_framework_llm_active,
    reset_framework_llm_active,
)
from openlit.semcov import SemanticConvention
from openlit.instrumentation.strands.utils import (
    emit_strands_inference_event,
    extract_content_from_events,
    infer_provider_name,
    infer_server_address,
    record_strands_metrics,
    truncate_content,
    truncate_message_content,
)

logger = logging.getLogger(__name__)

# The Strands SDK tracer uses ``__name__`` which resolves to this module path.
_STRANDS_TRACER_SCOPE = "strands.telemetry.tracer"


class StrandsSpanProcessor(SpanProcessor):
    """Enriches Strands-generated spans with OpenLIT telemetry."""

    def __init__(
        self,
        version,
        environment,
        application_name,
        capture_message_content,
        metrics,
        disable_metrics,
        event_provider=None,
        pricing_info=None,
    ):
        self._version = version
        self._environment = environment
        self._application_name = application_name
        self._capture_message_content = capture_message_content
        self._metrics = metrics
        self._disable_metrics = disable_metrics
        self._event_provider = event_provider
        self._pricing_info = pricing_info
        self._fw_tokens_lock = threading.Lock()
        self._fw_tokens = {}

    # -----------------------------------------------------------------
    # Span detection
    # -----------------------------------------------------------------

    @staticmethod
    def _is_strands_span(span):
        """Return ``True`` if *span* was created by the Strands SDK tracer."""
        scope = getattr(span, "instrumentation_scope", None)
        if scope and getattr(scope, "name", None) == _STRANDS_TRACER_SCOPE:
            return True
        attrs = getattr(span, "attributes", None) or {}
        return (
            attrs.get("gen_ai.system") == "strands-agents"
            or attrs.get("gen_ai.provider.name") == "strands-agents"
        )

    # -----------------------------------------------------------------
    # Private helpers
    # -----------------------------------------------------------------

    @staticmethod
    def _set_attr(span, key, value):
        """Safely set an attribute on a ``ReadableSpan`` via ``_attributes``.

        After ``on_end`` the span is read-only; the only way to enrich it
        is to mutate the private ``_attributes`` mapping (a
        ``MappingProxyType``).  This is the same pattern used by
        OpenInference and is wrapped in a try/except for safety.
        """
        try:
            raw = getattr(span, "_attributes", None)
            if raw is not None:
                new_attrs = dict(raw)
                new_attrs[key] = value
                span._attributes = types.MappingProxyType(new_attrs)
        except Exception:
            pass

    @staticmethod
    def _set_attrs(span, mapping):
        """Batch version of ``_set_attr``."""
        try:
            raw = getattr(span, "_attributes", None)
            if raw is not None:
                new_attrs = dict(raw)
                new_attrs.update(mapping)
                span._attributes = types.MappingProxyType(new_attrs)
        except Exception:
            pass

    # -----------------------------------------------------------------
    # SpanProcessor API
    # -----------------------------------------------------------------

    def on_start(self, span, parent_context=None):
        """Add static OpenLIT attributes while the span is still writable.

        For ``chat`` spans, sets ``_framework_llm_span_active`` so that
        provider-level instrumentors (OpenAI, Anthropic, ...) skip creating
        their own duplicate chat span.
        """
        if not self._is_strands_span(span):
            return

        try:
            span.set_attribute(TELEMETRY_SDK_NAME, "openlit")
            span.set_attribute(SemanticConvention.GEN_AI_SDK_VERSION, self._version)
            span.set_attribute(DEPLOYMENT_ENVIRONMENT, self._environment)
            span.set_attribute(SERVICE_NAME, self._application_name)
        except Exception:
            pass

        if getattr(span, "name", "") == "chat":
            try:
                token = set_framework_llm_active()
                span_id = span.get_span_context().span_id
                with self._fw_tokens_lock:
                    self._fw_tokens[span_id] = token
            except Exception:
                pass

    def on_end(self, span):
        """Enrich the finished span with derived attributes and record metrics."""
        if not self._is_strands_span(span):
            return

        span_id = getattr(getattr(span, "context", None), "span_id", None)
        if span_id is not None:
            with self._fw_tokens_lock:
                fw_token = self._fw_tokens.pop(span_id, None)
            if fw_token is not None:
                try:
                    reset_framework_llm_active(fw_token)
                except Exception:
                    pass

        try:
            self._process_span(span)
        except Exception:
            logger.debug("Error processing Strands span", exc_info=True)

    def shutdown(self):
        """No-op; nothing to clean up."""

    def force_flush(self, timeout_millis=None):
        """No-op."""

    # -----------------------------------------------------------------
    # Core processing
    # -----------------------------------------------------------------

    def _process_span(self, span):
        attrs = span.attributes or {}
        operation = attrs.get("gen_ai.operation.name", "")

        # Normalize gen_ai.system → gen_ai.provider.name
        gen_ai_system = attrs.get("gen_ai.system", "")
        if gen_ai_system and not attrs.get(SemanticConvention.GEN_AI_PROVIDER_NAME):
            self._set_attr(span, SemanticConvention.GEN_AI_PROVIDER_NAME, gen_ai_system)

        # Remap Strands-native system_prompt → gen_ai.system_instructions
        if operation == "invoke_agent":
            system_prompt = attrs.get("system_prompt")
            if system_prompt and not attrs.get(SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS):
                self._set_attr(
                    span, SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS, system_prompt
                )

        # Duration (nanoseconds → seconds)
        duration = 0.0
        if span.end_time and span.start_time:
            duration = (span.end_time - span.start_time) / 1e9
        self._set_attr(
            span, SemanticConvention.GEN_AI_CLIENT_OPERATION_DURATION, duration
        )

        # Server address / port (inferred from model name)
        model_name = attrs.get("gen_ai.request.model", "")
        server_address = attrs.get(SemanticConvention.SERVER_ADDRESS, "")
        server_port = attrs.get(SemanticConvention.SERVER_PORT, 0)
        if not server_address and model_name:
            server_address, server_port = infer_server_address(model_name)
            if server_address:
                self._set_attrs(span, {
                    SemanticConvention.SERVER_ADDRESS: server_address,
                    SemanticConvention.SERVER_PORT: server_port,
                })

        # Normalize multi-agent operation names to invoke_workflow
        if operation in ("invoke_swarm", "invoke_graph"):
            workflow_name = attrs.get("gen_ai.agent.name", "")
            self._set_attrs(span, {
                SemanticConvention.GEN_AI_OPERATION: SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
                SemanticConvention.GEN_AI_WORKFLOW_NAME: workflow_name,
            })
            operation = "invoke_workflow"

        # Output type for agent / workflow spans
        if operation in ("invoke_agent", "invoke_workflow"):
            self._set_attr(
                span,
                SemanticConvention.GEN_AI_OUTPUT_TYPE,
                SemanticConvention.GEN_AI_OUTPUT_TYPE_TEXT,
            )

        # Tool type
        if operation == "execute_tool":
            self._set_attr(span, SemanticConvention.GEN_AI_TOOL_TYPE, "function")

        # Chat span enrichment: match provider span attributes
        if operation == "chat":
            self._enrich_chat_span(span, attrs, model_name)

        # Content capture: extract from events → span attributes
        if self._capture_message_content:
            self._extract_and_set_content(span, operation)

        # Emit inference log event for chat spans (provider span is
        # suppressed, so this processor is the single source of truth)
        if operation == "chat" and self._event_provider:
            self._emit_chat_inference_event(span, attrs, server_address, server_port)

        # Record OpenLIT metrics (skip spans without a recognized operation)
        if not self._disable_metrics and self._metrics and operation:
            record_strands_metrics(
                self._metrics,
                operation,
                duration,
                self._environment,
                self._application_name,
                model_name,
                server_address,
                server_port,
            )

    # -----------------------------------------------------------------
    # Chat span enrichment (parity with provider spans)
    # -----------------------------------------------------------------

    def _enrich_chat_span(self, span, attrs, model_name):
        """Add attributes to chat spans that match provider-level chat spans."""
        enrichments = {}

        # Span name: "chat" → "chat {model}" (matching OpenAI pattern)
        if model_name:
            try:
                span._name = f"chat {model_name}"
            except Exception:
                pass

        # Override gen_ai.provider.name with actual provider for chat spans
        provider = infer_provider_name(model_name) if model_name else ""
        if provider:
            enrichments[SemanticConvention.GEN_AI_PROVIDER_NAME] = provider

        # Finish reasons from output events
        _, output_msgs, _ = extract_content_from_events(span, "chat")
        if output_msgs:
            finish_reasons = [
                m.get("finish_reason")
                for m in output_msgs
                if isinstance(m, dict) and m.get("finish_reason")
            ]
            if finish_reasons:
                enrichments[SemanticConvention.GEN_AI_RESPONSE_FINISH_REASON] = finish_reasons

        enrichments[SemanticConvention.GEN_AI_OUTPUT_TYPE] = (
            SemanticConvention.GEN_AI_OUTPUT_TYPE_TEXT
        )
        enrichments[SemanticConvention.GEN_AI_REQUEST_IS_STREAM] = True

        # Token totals and cost
        input_tokens = attrs.get("gen_ai.usage.input_tokens", 0)
        output_tokens = attrs.get("gen_ai.usage.output_tokens", 0)
        if input_tokens or output_tokens:
            enrichments[SemanticConvention.GEN_AI_CLIENT_TOKEN_USAGE] = (
                int(input_tokens) + int(output_tokens)
            )
        if self._pricing_info and model_name:
            cost = get_chat_model_cost(
                model_name, self._pricing_info, input_tokens, output_tokens
            )
            enrichments[SemanticConvention.GEN_AI_USAGE_COST] = cost

        if enrichments:
            self._set_attrs(span, enrichments)

    # -----------------------------------------------------------------
    # Content extraction → span attributes
    # -----------------------------------------------------------------

    def _extract_and_set_content(self, span, operation):
        """Extract messages from span events and set as span attributes."""
        try:
            input_msgs, output_msgs, system_instr = extract_content_from_events(
                span, operation
            )

            additions = {}

            if operation == "execute_tool":
                # For tool spans: map to tool-specific attributes
                if input_msgs:
                    first = input_msgs[0] if isinstance(input_msgs, list) else input_msgs
                    parts = first.get("parts", []) if isinstance(first, dict) else []
                    if parts:
                        arguments = parts[0].get("arguments", parts[0].get("response", ""))
                        additions[SemanticConvention.GEN_AI_TOOL_CALL_ARGUMENTS] = (
                            truncate_content(
                                json.dumps(arguments) if not isinstance(arguments, str) else arguments
                            )
                        )
                if output_msgs:
                    additions[SemanticConvention.GEN_AI_TOOL_CALL_RESULT] = (
                        truncate_content(json.dumps(output_msgs))
                    )
            else:
                if input_msgs:
                    truncate_message_content(input_msgs)
                    additions[SemanticConvention.GEN_AI_INPUT_MESSAGES] = json.dumps(
                        input_msgs
                    )
                if output_msgs:
                    truncate_message_content(output_msgs)
                    additions[SemanticConvention.GEN_AI_OUTPUT_MESSAGES] = json.dumps(
                        output_msgs
                    )
                if system_instr:
                    additions[SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS] = (
                        truncate_content(str(system_instr))
                    )

            if additions:
                self._set_attrs(span, additions)
        except Exception:
            logger.debug("Error extracting Strands content", exc_info=True)

    # -----------------------------------------------------------------
    # Chat inference log event
    # -----------------------------------------------------------------

    def _emit_chat_inference_event(self, span, attrs, server_address, server_port):
        """Emit ``gen_ai.client.inference.operation.details`` for chat spans."""
        try:
            input_msgs, output_msgs, system_instr = extract_content_from_events(
                span, "chat"
            )

            extra = {}

            input_tokens = attrs.get("gen_ai.usage.input_tokens")
            output_tokens = attrs.get("gen_ai.usage.output_tokens")
            if input_tokens is not None:
                extra["input_tokens"] = input_tokens
            if output_tokens is not None:
                extra["output_tokens"] = output_tokens

            cache_read = attrs.get("gen_ai.usage.cache_read_input_tokens")
            cache_write = attrs.get("gen_ai.usage.cache_write_input_tokens")
            if cache_read is not None:
                extra["cache_read_input_tokens"] = cache_read
            if cache_write is not None:
                extra["cache_creation_input_tokens"] = cache_write

            if system_instr:
                extra["system_instructions"] = system_instr

            # Finish reason from output messages
            if output_msgs:
                finish_reasons = [
                    m.get("finish_reason")
                    for m in output_msgs
                    if isinstance(m, dict) and m.get("finish_reason")
                ]
                if finish_reasons:
                    extra["finish_reasons"] = finish_reasons

            emit_strands_inference_event(
                event_provider=self._event_provider,
                operation_name=SemanticConvention.GEN_AI_OPERATION_TYPE_CHAT,
                request_model=attrs.get("gen_ai.request.model"),
                input_messages=input_msgs if self._capture_message_content else None,
                output_messages=output_msgs if self._capture_message_content else None,
                server_address=server_address,
                server_port=server_port,
                **extra,
            )
        except Exception:
            logger.debug("Error emitting Strands inference event", exc_info=True)

File: sdk/typescript/src/instrumentation/strands/processor.ts
/**
 * Strands Agents SpanProcessor.
 *
 * Enriches Strands' native OTel spans with OpenLIT-specific attributes,
 * extracts content from span events into span attributes, emits
 * gen_ai.client.inference.operation.details log events for chat spans,
 * and records OpenLIT metrics.
 *
 * Provider-level chat spans (OpenAI, Anthropic, etc.) are suppressed
 * when they occur inside a Strands chat span via the shared
 * frameworkLlmActive flag.
 *
 * Mirrors: sdk/python/src/openlit/instrumentation/strands/processor.py
 */

import { Context, SpanKind, SpanStatusCode } from '@opentelemetry/api';
import { ATTR_SERVICE_NAME, ATTR_TELEMETRY_SDK_NAME } from '@opentelemetry/semantic-conventions';
import type { SpanProcessor, ReadableSpan } from '@opentelemetry/sdk-trace-base';
import type { Span } from '@opentelemetry/api';

import SemanticConvention from '../../semantic-convention';
import OpenlitConfig from '../../config';
import OpenLitHelper, {
  applyCustomSpanAttributes,
  setFrameworkLlmActive,
  resetFrameworkLlmActive,
} from '../../helpers';
import {
  extractContentFromEvents,
  inferProviderName,
  inferServerAddress,
  truncateContent,
  truncateMessageContent,
  recordStrandsMetrics,
  emitStrandsInferenceEvent,
} from './utils';

const STRANDS_TRACER_SCOPE = 'strands.telemetry.tracer';

/**
 * Enriches Strands-generated spans with OpenLIT telemetry.
 * Added to the TracerProvider so it receives all spans; non-Strands
 * spans are ignored via the _isStrandsSpan() check.
 */
export class StrandsSpanProcessor implements SpanProcessor {
  private _strandsVersion: string;
  private _chatSpanIds = new Set<string>();
  private _chatInfo = new Map<string, Record<string, any>>();

  constructor(strandsVersion: string = 'unknown') {
    this._strandsVersion = strandsVersion;
  }

  // -----------------------------------------------------------------
  // Span detection
  // -----------------------------------------------------------------

  private static _isStrandsSpan(span: any): boolean {
    const scope = span.instrumentationLibrary;
    if (scope && scope.name === STRANDS_TRACER_SCOPE) {
      return true;
    }
    const attrs = span.attributes || {};
    return (
      attrs['gen_ai.system'] === SemanticConvention.GEN_AI_SYSTEM_STRANDS ||
      attrs['gen_ai.provider.name'] === SemanticConvention.GEN_AI_SYSTEM_STRANDS
    );
  }

  // -----------------------------------------------------------------
  // Attribute mutation helpers (span is read-only after onEnd)
  // -----------------------------------------------------------------

  private static _setAttr(span: any, key: string, value: any): void {
    try {
      if (span.attributes) {
        span.attributes[key] = value;
      }
    } catch {
      // ignore
    }
  }

  private static _setAttrs(span: any, mapping: Record<string, any>): void {
    try {
      if (span.attributes) {
        Object.assign(span.attributes, mapping);
      }
    } catch {
      // ignore
    }
  }

  // -----------------------------------------------------------------
  // SpanProcessor API
  // -----------------------------------------------------------------

  onStart(span: Span, _parentContext: Context): void {
    if (!StrandsSpanProcessor._isStrandsSpan(span as any)) return;

    try {
      span.setAttribute(ATTR_TELEMETRY_SDK_NAME, 'openlit');
      span.setAttribute(SemanticConvention.GEN_AI_SDK_VERSION, this._strandsVersion);
      span.setAttribute(
        SemanticConvention.ATTR_DEPLOYMENT_ENVIRONMENT,
        OpenlitConfig.environment || 'default',
      );
      span.setAttribute(ATTR_SERVICE_NAME, OpenlitConfig.applicationName || 'default');
      applyCustomSpanAttributes(span);
    } catch {
      // ignore
    }

    const spanName = (span as any).name || '';
    if (spanName === 'chat') {
      try {
        (span as any)._kind = SpanKind.CLIENT;
      } catch {
        // ignore
      }
      try {
        setFrameworkLlmActive();
        const spanId = span.spanContext().spanId;
        this._chatSpanIds.add(spanId);
      } catch {
        // ignore
      }
    }
  }

  onEnd(span: ReadableSpan): void {
    if (!StrandsSpanProcessor._isStrandsSpan(span)) return;

    const spanId = span.spanContext().spanId;
    if (this._chatSpanIds.has(spanId)) {
      this._chatSpanIds.delete(spanId);
      try {
        resetFrameworkLlmActive();
      } catch {
        // ignore
      }
    }

    try {
      this._processSpan(span);
    } catch {
      // ignore
    }
  }

  shutdown(): Promise<void> {
    return Promise.resolve();
  }

  forceFlush(): Promise<void> {
    return Promise.resolve();
  }

  // -----------------------------------------------------------------
  // Core processing (mirrors Python _process_span)
  // -----------------------------------------------------------------

  private _processSpan(span: ReadableSpan): void {
    const attrs = span.attributes || {};
    let operation = String(attrs[SemanticConvention.GEN_AI_OPERATION] || '');

    // Normalize agent id: agent_name-span_id_hex
    const agentName = attrs[SemanticConvention.GEN_AI_AGENT_NAME] as string | undefined;
    if (agentName && !attrs[SemanticConvention.GEN_AI_AGENT_ID]) {
      const spanIdHex = span.spanContext().spanId;
      StrandsSpanProcessor._setAttr(
        span,
        SemanticConvention.GEN_AI_AGENT_ID,
        `${agentName}-${spanIdHex}`,
      );
    }

    // Normalize gen_ai.system → gen_ai.provider.name
    const genAiSystem = String(attrs['gen_ai.system'] || '');
    if (genAiSystem && !attrs[SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL]) {
      const provider =
        genAiSystem === 'strands-agents' || genAiSystem === 'strands_agents'
          ? SemanticConvention.GEN_AI_SYSTEM_STRANDS
          : genAiSystem;
      StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL, provider);
    }

    // Normalize Strands-native cache token keys → OTel standard keys
    const cacheKeyMap: [string, string][] = [
      ['gen_ai.usage.cache_read_input_tokens', SemanticConvention.GEN_AI_USAGE_CACHE_READ_INPUT_TOKENS],
      ['gen_ai.usage.cache_write_input_tokens', SemanticConvention.GEN_AI_USAGE_CACHE_CREATION_INPUT_TOKENS],
    ];
    for (const [strandsKey, otelKey] of cacheKeyMap) {
      const val = attrs[strandsKey];
      if (val != null && !attrs[otelKey]) {
        StrandsSpanProcessor._setAttr(span, otelKey, val);
      }
    }

    // Remap Strands-native system_prompt → gen_ai.system_instructions
    if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_AGENT) {
      const systemPrompt = attrs['system_prompt'] as string | undefined;
      if (systemPrompt && !attrs[SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS]) {
        StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS, systemPrompt);
      }
    }

    // Duration (HrTime → seconds)
    let duration = 0;
    if (span.endTime && span.startTime) {
      const endNs = span.endTime[0] * 1e9 + span.endTime[1];
      const startNs = span.startTime[0] * 1e9 + span.startTime[1];
      duration = (endNs - startNs) / 1e9;
    }
    StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_CLIENT_OPERATION_DURATION, duration);

    // Server address / port (inferred from model name)
    const modelName = String(attrs[SemanticConvention.GEN_AI_REQUEST_MODEL] || '');
    let serverAddress = String(attrs[SemanticConvention.SERVER_ADDRESS] || '');
    let serverPort = Number(attrs[SemanticConvention.SERVER_PORT] || 0);
    if (!serverAddress && modelName) {
      [serverAddress, serverPort] = inferServerAddress(modelName);
      if (serverAddress) {
        StrandsSpanProcessor._setAttrs(span, {
          [SemanticConvention.SERVER_ADDRESS]: serverAddress,
          [SemanticConvention.SERVER_PORT]: serverPort,
        });
      }
    }

    // Normalize multi-agent operation names to invoke_workflow
    if (operation === 'invoke_swarm' || operation === 'invoke_graph') {
      const workflowName = String(attrs[SemanticConvention.GEN_AI_AGENT_NAME] || '');
      StrandsSpanProcessor._setAttrs(span, {
        [SemanticConvention.GEN_AI_OPERATION]: SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
        [SemanticConvention.GEN_AI_WORKFLOW_NAME]: workflowName,
      });
      operation = SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK;
    }

    // Output type for agent / workflow spans
    if (
      operation === SemanticConvention.GEN_AI_OPERATION_TYPE_AGENT ||
      operation === SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK
    ) {
      StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_OUTPUT_TYPE, SemanticConvention.GEN_AI_OUTPUT_TYPE_TEXT);
    }

    // Tool type and tool call id
    if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_TOOLS) {
      StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_TOOL_TYPE, 'function');
      if (!attrs[SemanticConvention.GEN_AI_TOOL_CALL_ID]) {
        const tid =
          attrs['tool_use_id'] ||
          attrs['toolUseId'] ||
          attrs['gen_ai.tool.call.id'] ||
          StrandsSpanProcessor._extractToolCallIdFromSpanEvents(span);
        if (tid) {
          StrandsSpanProcessor._setAttr(span, SemanticConvention.GEN_AI_TOOL_CALL_ID, String(tid));
        }
      }
    }

    // OTel-compliant span names
    this._setOtelCompliantSpanName(span, operation);

    // Chat span enrichment: match provider span attributes
    if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_CHAT) {
      this._enrichChatSpan(span, attrs, modelName);
      this._storeChatInfoForParent(span, modelName);
    }

    // Propagate recommended attrs from child chat spans to invoke_agent
    if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_AGENT) {
      this._enrichAgentFromChildren(span);
    }

    // Content capture: extract from events → span attributes
    if (OpenlitConfig.captureMessageContent) {
      this._extractAndSetContent(span, operation);
    }

    // Emit inference log event for chat spans
    if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_CHAT) {
      this._emitChatInferenceEvent(span, attrs, serverAddress, serverPort);
    }

    // Record OpenLIT metrics
    if (!OpenlitConfig.disableMetrics && operation) {
      recordStrandsMetrics(operation, duration, modelName, serverAddress, serverPort);
    }

    // Set error type if missing (low-cardinality per OTel spec)
    if (span.status && span.status.code === SpanStatusCode.ERROR) {
      const currentAttrs = span.attributes || {};
      if (!currentAttrs[SemanticConvention.ERROR_TYPE]) {
        StrandsSpanProcessor._setAttr(
          span,
          SemanticConvention.ERROR_TYPE,
          '_OTHER',
        );
      }
    }
  }

  // -----------------------------------------------------------------
  // OTel-compliant span naming
  // -----------------------------------------------------------------

  private _setOtelCompliantSpanName(span: ReadableSpan, operation: string): void {
    if (
      operation !== SemanticConvention.GEN_AI_OPERATION_TYPE_AGENT &&
      operation !== SemanticConvention.GEN_AI_OPERATION_TYPE_TOOLS &&
      operation !== SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK
    ) {
      return;
    }
    try {
      const attrs = span.attributes || {};
      if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_AGENT) {
        const name = attrs[SemanticConvention.GEN_AI_AGENT_NAME];
        if (name) (span as any)._name = `invoke_agent ${name}`;
      } else if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_TOOLS) {
        const name = attrs[SemanticConvention.GEN_AI_TOOL_NAME];
        if (name) (span as any)._name = `execute_tool ${name}`;
      } else if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK) {
        const name = attrs[SemanticConvention.GEN_AI_WORKFLOW_NAME];
        if (name) (span as any)._name = `invoke_workflow ${name}`;
      }
    } catch {
      // ignore
    }
  }

  // -----------------------------------------------------------------
  // Chat span enrichment (parity with provider spans)
  // -----------------------------------------------------------------

  private _enrichChatSpan(span: ReadableSpan, attrs: Record<string, any>, modelName: string): void {
    const enrichments: Record<string, any> = {};

    // Span name: "chat" → "chat {model}"
    if (modelName) {
      try {
        (span as any)._name = `chat ${modelName}`;
      } catch {
        // ignore
      }
    }

    // Override gen_ai.provider.name with actual provider for chat spans
    const provider = modelName ? inferProviderName(modelName) : '';
    if (provider) {
      enrichments[SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL] = provider;
    }

    // response.model: fall back to request model
    if (!attrs[SemanticConvention.GEN_AI_RESPONSE_MODEL] && modelName) {
      enrichments[SemanticConvention.GEN_AI_RESPONSE_MODEL] = modelName;
    }

    // response.id: extract from events
    if (!attrs[SemanticConvention.GEN_AI_RESPONSE_ID]) {
      const responseId = StrandsSpanProcessor._extractResponseId(span);
      if (responseId) {
        enrichments[SemanticConvention.GEN_AI_RESPONSE_ID] = responseId;
      }
    }

    // Finish reasons from output events
    const [, outputMsgs] = extractContentFromEvents(span, 'chat');
    if (outputMsgs && outputMsgs.length > 0) {
      const finishReasons = outputMsgs
        .filter((m: any) => typeof m === 'object' && m.finish_reason)
        .map((m: any) => m.finish_reason);
      if (finishReasons.length > 0) {
        enrichments[SemanticConvention.GEN_AI_RESPONSE_FINISH_REASON] = finishReasons;
      }
    }

    enrichments[SemanticConvention.GEN_AI_OUTPUT_TYPE] = SemanticConvention.GEN_AI_OUTPUT_TYPE_TEXT;

    // Token totals and cost
    const inputTokens = Number(attrs['gen_ai.usage.input_tokens'] || 0);
    const outputTokens = Number(attrs['gen_ai.usage.output_tokens'] || 0);
    if (inputTokens || outputTokens) {
      enrichments[SemanticConvention.GEN_AI_CLIENT_TOKEN_USAGE] = inputTokens + outputTokens;
    }
    if (OpenlitConfig.pricingInfo && modelName) {
      const cost = OpenLitHelper.getChatModelCost(modelName, OpenlitConfig.pricingInfo, inputTokens, outputTokens);
      enrichments[SemanticConvention.GEN_AI_USAGE_COST] = cost;
    }

    if (Object.keys(enrichments).length > 0) {
      StrandsSpanProcessor._setAttrs(span, enrichments);
    }
  }

  private _storeChatInfoForParent(span: ReadableSpan, modelName: string): void {
    try {
      const parentId = (span as any).parentSpanId;
      if (!parentId) return;

      const finalAttrs = span.attributes || {};
      const info: Record<string, any> = {
        responseModel: finalAttrs[SemanticConvention.GEN_AI_RESPONSE_MODEL] || modelName,
        responseId: finalAttrs[SemanticConvention.GEN_AI_RESPONSE_ID],
        finishReasons: finalAttrs[SemanticConvention.GEN_AI_RESPONSE_FINISH_REASON],
        inputTokens: finalAttrs['gen_ai.usage.input_tokens'] || 0,
        outputTokens: finalAttrs['gen_ai.usage.output_tokens'] || 0,
      };
      this._chatInfo.set(parentId, info);
    } catch {
      // ignore
    }
  }

  private _enrichAgentFromChildren(span: ReadableSpan): void {
    try {
      const spanId = span.spanContext().spanId;
      const info = this._chatInfo.get(spanId);
      this._chatInfo.delete(spanId);
      if (!info) return;

      const enrichments: Record<string, any> = {};
      const current = span.attributes || {};

      if (info.responseModel && !current[SemanticConvention.GEN_AI_RESPONSE_MODEL]) {
        enrichments[SemanticConvention.GEN_AI_RESPONSE_MODEL] = info.responseModel;
      }
      if (info.responseId && !current[SemanticConvention.GEN_AI_RESPONSE_ID]) {
        enrichments[SemanticConvention.GEN_AI_RESPONSE_ID] = info.responseId;
      }
      if (info.finishReasons && !current[SemanticConvention.GEN_AI_RESPONSE_FINISH_REASON]) {
        enrichments[SemanticConvention.GEN_AI_RESPONSE_FINISH_REASON] = info.finishReasons;
      }
      if (Object.keys(enrichments).length > 0) {
        StrandsSpanProcessor._setAttrs(span, enrichments);
      }
    } catch {
      // ignore
    }
  }

  // -----------------------------------------------------------------
  // Static extraction helpers
  // -----------------------------------------------------------------

  private static _extractResponseId(span: ReadableSpan): string {
    for (const event of span.events || []) {
      const ea = event.attributes || {};
      const rid = ea['gen_ai.response.id'] || ea['response_id'];
      if (rid) return String(rid);
    }
    return '';
  }

  private static _extractToolCallIdFromSpanEvents(span: ReadableSpan): string | null {
    for (const event of span.events || []) {
      if (event.name === 'gen_ai.tool.message') {
        const ea = event.attributes || {};
        const tid = ea['id'] || ea[SemanticConvention.GEN_AI_TOOL_CALL_ID];
        if (tid) return String(tid);
      }
    }
    for (const event of span.events || []) {
      const ea = event.attributes || {};
      const tid =
        ea[SemanticConvention.GEN_AI_TOOL_CALL_ID] ||
        ea['tool_use_id'] ||
        ea['toolUseId'] ||
        ea['gen_ai.tool.call.id'];
      if (tid) return String(tid);
    }
    return null;
  }

  // -----------------------------------------------------------------
  // Content extraction → span attributes
  // -----------------------------------------------------------------

  private _extractAndSetContent(span: ReadableSpan, operation: string): void {
    try {
      const [inputMsgs, outputMsgs, systemInstr] = extractContentFromEvents(span, operation);
      const additions: Record<string, any> = {};

      if (operation === SemanticConvention.GEN_AI_OPERATION_TYPE_TOOLS) {
        if (inputMsgs.length > 0) {
          const first = inputMsgs[0];
          const parts = (typeof first === 'object' && first.parts) ? first.parts : [];
          if (parts.length > 0) {
            const arguments_ = parts[0].arguments || parts[0].response || '';
            additions[SemanticConvention.GEN_AI_TOOL_CALL_ARGUMENTS] = truncateContent(
              typeof arguments_ === 'string' ? arguments_ : JSON.stringify(arguments_),
            );
          }
        }
        if (outputMsgs.length > 0) {
          additions[SemanticConvention.GEN_AI_TOOL_CALL_RESULT] = truncateContent(
            JSON.stringify(outputMsgs),
          );
        }
      } else {
        if (inputMsgs.length > 0) {
          truncateMessageContent(inputMsgs);
          additions[SemanticConvention.GEN_AI_INPUT_MESSAGES] = JSON.stringify(inputMsgs);
        }
        if (outputMsgs.length > 0) {
          truncateMessageContent(outputMsgs);
          additions[SemanticConvention.GEN_AI_OUTPUT_MESSAGES] = JSON.stringify(outputMsgs);
        }
        if (systemInstr) {
          additions[SemanticConvention.GEN_AI_SYSTEM_INSTRUCTIONS] = truncateContent(String(systemInstr));
        }
      }

      if (Object.keys(additions).length > 0) {
        StrandsSpanProcessor._setAttrs(span, additions);
      }
    } catch {
      // ignore
    }
  }

  // -----------------------------------------------------------------
  // Chat inference log event
  // -----------------------------------------------------------------

  private _emitChatInferenceEvent(
    span: ReadableSpan,
    attrs: Record<string, any>,
    serverAddress: string,
    serverPort: number,
  ): void {
    try {
      const [inputMsgs, outputMsgs, systemInstr] = extractContentFromEvents(span, 'chat');

      const extra: Record<string, any> = {};

      const inputTokens = attrs['gen_ai.usage.input_tokens'];
      const outputTokens = attrs['gen_ai.usage.output_tokens'];
      if (inputTokens != null) extra.inputTokens = inputTokens;
      if (outputTokens != null) extra.outputTokens = outputTokens;

      const cacheRead =
        attrs[SemanticConvention.GEN_AI_USAGE_CACHE_READ_INPUT_TOKENS] ||
        attrs['gen_ai.usage.cache_read_input_tokens'];
      const cacheWrite =
        attrs[SemanticConvention.GEN_AI_USAGE_CACHE_CREATION_INPUT_TOKENS] ||
        attrs['gen_ai.usage.cache_write_input_tokens'];
      if (cacheRead != null) extra.cacheReadInputTokens = cacheRead;
      if (cacheWrite != null) extra.cacheCreationInputTokens = cacheWrite;

      const responseId =
        attrs[SemanticConvention.GEN_AI_RESPONSE_ID] ||
        StrandsSpanProcessor._extractResponseId(span);
      if (responseId) extra.responseId = responseId;

      if (systemInstr) extra.systemInstructions = systemInstr;

      if (outputMsgs.length > 0) {
        const finishReasons = outputMsgs
          .filter((m: any) => typeof m === 'object' && m.finish_reason)
          .map((m: any) => m.finish_reason);
        if (finishReasons.length > 0) extra.finishReasons = finishReasons;
      }

      extra.inputMessages = inputMsgs;
      extra.outputMessages = outputMsgs;

      emitStrandsInferenceEvent(
        span,
        String(attrs[SemanticConvention.GEN_AI_REQUEST_MODEL] || ''),
        serverAddress,
        serverPort,
        extra,
      );
    } catch {
      // ignore
    }
  }
}
